modules/tensor_mechanics/src/materials/czm/CZMMaterialBase.C

```diff
@@ -0,0 +1,79 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "Assembly.h"
+#include "CZMMaterialBase.h"
+#include "RotationMatrix.h"
+
+template <>
+InputParameters
+validParams<CZMMaterialBase>()
+{
+  InputParameters params = validParams<InterfaceMaterial>();
+
+  params.addClassDescription("Base class for cohesive zone mateirla models");
+  params.addRequiredCoupledVar("displacements",
+                               "The string of displacements suitable for the problem statement");
+  return params;
+}
+
+CZMMaterialBase::CZMMaterialBase(const InputParameters & parameters)
+  : InterfaceMaterial(parameters),
+    _normals(_assembly.normals()),
+    _ndisp(coupledComponents("displacements")),
+    _disp(_ndisp),
+    _disp_neighbor(_ndisp),
+    _displacement_jump_global(declareProperty<RealVectorValue>("displacement_jump_global")),
+    _displacement_jump(declareProperty<RealVectorValue>("displacement_jump")),
+    _traction_global(declareProperty<RealVectorValue>("traction_global")),
+    _traction(declareProperty<RealVectorValue>("traction")),
+    _traction_derivatives_global(declareProperty<RankTwoTensor>("traction_derivatives_global")),
+    _traction_derivatives(declareProperty<RankTwoTensor>("traction_derivatives"))
+{
+  if (_ndisp > 3 || _ndisp < 1)
+    mooseError("the CZM material requires 1, 2 or 3 displacement variables");
+
+  if (getParam<bool>("use_displaced_mesh") == true)
+    mooseError("This material cannot be used with use_displaced_mesh = true");
+
+  // initializing the displacement vectors
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
+    _disp[i] = &coupledValue("displacements", i);
+    _disp_neighbor[i] = &coupledNeighborValue("displacements", i);
+  }
+}
+
+void
+CZMMaterialBase::computeQpProperties()
+{
+
+  RealTensorValue RotationGlobalToLocal =
+      RotationMatrix::rotVec1ToVec2(_normals[_qp], RealVectorValue(1, 0, 0));
```

Comment:
@arovinelli this does not seem to be correct to me. RotationMatrix says "provides a rotation matrix that will rotate the vector vec to the z axis (the "2" direction)". For 2D, it seems to rotate a vector (x,y,0) to (x',0,y').

Reply:
> @arovinelli this does not seem to be correct to me. RotationMatrix says "provides a rotation matrix that will rotate the vector vec to the z axis (the "2" direction)".

You are referring to [rotVecToZ](https://github.com/idaholab/moose/blob/d79b9185bcd10b1a9e19c3b0b01ca46bd5794c52/framework/src/utils/RotationMatrix.C#L13)

but we are using [rotVec1ToVec2](https://github.com/idaholab/moose/blob/d79b9185bcd10b1a9e19c3b0b01ca46bd5794c52/framework/src/utils/RotationMatrix.C#L51) to rotate the normal to be (1,0,0) in the interface space.

If you rotate the normal, it might happened to be (0.99999, some small, some small) because of tolerances, but should always be something like that.
Then the two tangential direction are not guaranteed to be (0,1,0) and (0,0,1) in the interface space, they can be any direction in (0,y',z') as long as they are orthogonal. 
This is why the material should always be coded assuming 3 directions and not 2
hope this make sense

Reply:
I see your point! Thanks.